helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java

```diff
@@ -0,0 +1,82 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import com.google.common.collect.ImmutableMap;
+import org.apache.helix.HelixAdmin;
+import org.apache.helix.HelixDataAccessor;
+import org.apache.helix.HelixException;
+import org.apache.helix.PropertyKey;
+import org.apache.helix.TestHelper;
+import org.apache.helix.task.JobConfig;
+import org.apache.helix.task.JobQueue;
+import org.apache.helix.task.TaskState;
+import org.apache.helix.task.TaskUtil;
+import org.apache.helix.task.Workflow;
+import org.apache.helix.task.WorkflowContext;
+import org.apache.helix.tools.ClusterVerifiers.BestPossibleExternalViewVerifier;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+/**
+ * Test To check if the Execution Delay is respected.
+ * If Workflow is completed regardless of execution delay, it means this task should be failed.
+ * If execution delay is respected, the task will be passed.
+ */
+public class TestExecutionDelay extends TaskTestBase {
+
+  @BeforeClass
+  public void beforeClass() throws Exception {
+    super.beforeClass();
+  }
+
+  @Test(expectedExceptions = HelixException.class)
+  public void testExecutionDelay() throws InterruptedException {
+    final long timeOutTime = 15000L;
+    final long executionDelay = 100000000L;
+    String workflowName = TestHelper.getTestMethodName();
+    Workflow.Builder builder = new Workflow.Builder(workflowName);
+
+    JobConfig.Builder jobBuilder = JobConfig.Builder.fromMap(WorkflowGenerator.DEFAULT_JOB_CONFIG)
+        .setMaxAttemptsPerTask(1).setWorkflow(workflowName)
+        .setJobCommandConfigMap(ImmutableMap.of(MockTask.JOB_DELAY, "1000"));
+
+    JobConfig.Builder jobBuilder2 = JobConfig.Builder.fromMap(WorkflowGenerator.DEFAULT_JOB_CONFIG)
+        .setMaxAttemptsPerTask(1).setWorkflow(workflowName)
+        .setJobCommandConfigMap(ImmutableMap.of(MockTask.JOB_DELAY, "1000"));
+
+    JobConfig.Builder jobBuilder3 = JobConfig.Builder.fromMap(WorkflowGenerator.DEFAULT_JOB_CONFIG)
+        .setMaxAttemptsPerTask(1).setWorkflow(workflowName)
+        .setJobCommandConfigMap(ImmutableMap.of(MockTask.JOB_DELAY, "1000"));
+
+    builder.addParentChildDependency("JOB0", "JOB1");
+    builder.addParentChildDependency("JOB0", "JOB2");
+    builder.addJob("JOB0", jobBuilder.setExecutionDelay(executionDelay));
+    builder.addJob("JOB1", jobBuilder2);
+    builder.addJob("JOB2", jobBuilder3);
+
+    // Since the execution delay is set as a long time, it is expected to reach time out.
+    // If the workflow completed without any exception, it means that ExecutionDelay has not been
+    // respected.
+    _driver.start(builder.build());
+    _driver.pollForWorkflowState(workflowName, timeOutTime, TaskState.COMPLETED);
```

Comment:
I think there's a better way to test this. Instead of relying on pollForWorkflowState timing out, could you check the status of the workflow **and** the individual jobs? What if the controller is very slow and takes 15 seconds to schedule jobs?

One way of not relying on timing is
1. Start workflow
2. Wait until it is in progress and job 0 has been looked at (controller tried to schedule)
3. Read the context and check start time is set with the right amount
4. Check that job 0 and its children jobs have not run by checking their states

This way, your test is robust against timing issues, correct? No matter how long controller takes to start and schedule that first job instead of relying on 15 sec, an arbitrarily long enough number? :)